Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.mm

```diff
@@ -202,11 +202,37 @@ - (void)outputVideoEffectDidStopForStream:(SCStream *)stream
         ScreenCaptureKitSharingSessionManager::singleton().cancelPendingSessionForDevice(m_captureDevice);
 
     clearSharingSession();
+
+    if (m_whenReadyCallback)
+        std::exchange(m_whenReadyCallback, { })({ "Source no longer needed"_s , MediaAccessDenialReason::InvalidAccess });
+}
+
+void ScreenCaptureKitCaptureSource::whenReady(CompletionHandler<void(CaptureSourceError&&)>&& callback)
+{
+    if (m_didReceiveVideoFrame) {
+        callback({ });
+        return;
+    }
+
+    if (m_isRunning) {
+        m_whenReadyCallback = WTFMove(callback);
+        return;
+    }
+
+    m_whenReadyCallback = [weakThis = WeakPtr { *this }, callback = WTFMove(callback)] (auto&& result) mutable {
+        callback(WTFMove(result));
+        if (weakThis)
+            weakThis->stop();
+    };
+
+    // We start to get the first frame and stop after that point.
+    start();
 }
 
 bool ScreenCaptureKitCaptureSource::start()
 {
     ASSERT(isAvailable());
+    ASSERT(!m_whenReadyCallback);
```

Comment:
Won't we always hit this assert when `ScreenCaptureKitCaptureSource::whenReady` sets `m_whenReadyCallback` and then calls `start()`?

Reply:
Right, I updated the ASSERT accordingly.